web_src/js/features/user-auth-webauthn.js

```diff
@@ -180,40 +199,52 @@ export function initUserAuthWebAuthnRegister() {
   });
 }
 
-function webAuthnRegisterRequest() {
-  if ($('#nickname').val() === '') {
-    webAuthnError('empty');
+async function webAuthnRegisterRequest() {
+  const elNickname = document.getElementById('nickname');
+
+  const body = new FormData();
+  body.append('name', elNickname.value);
+
+  const resp = await fetch(`${appSubUrl}/user/settings/security/webauthn/request_register`, {
+    method: 'POST',
+    headers: {
+      'X-Csrf-Token': csrfToken,
+    },
+    body,
+  });
+
+  if (resp.status === 409) {
+    webAuthnError('duplicated');
+    return;
+  } else if (resp.status !== 200) {
+    webAuthnError('unknown');
     return;
   }
-  $.post(`${appSubUrl}/user/settings/security/webauthn/request_register`, {
-    _csrf: csrfToken,
-    name: $('#nickname').val(),
-  }).done((makeCredentialOptions) => {
-    $('#nickname').closest('div.field').removeClass('error');
-
-    makeCredentialOptions.publicKey.challenge = decodeURLEncodedBase64(makeCredentialOptions.publicKey.challenge);
-    makeCredentialOptions.publicKey.user.id = decodeURLEncodedBase64(makeCredentialOptions.publicKey.user.id);
-    if (makeCredentialOptions.publicKey.excludeCredentials) {
-      for (let i = 0; i < makeCredentialOptions.publicKey.excludeCredentials.length; i++) {
-        makeCredentialOptions.publicKey.excludeCredentials[i].id = decodeURLEncodedBase64(makeCredentialOptions.publicKey.excludeCredentials[i].id);
-      }
+
+  const options = await resp.json();
+  elNickname.closest('div.field').classList.remove('error');
+
+  options.publicKey.challenge = decodeURLEncodedBase64(options.publicKey.challenge);
+  options.publicKey.user.id = decodeURLEncodedBase64(options.publicKey.user.id);
+  if (options.publicKey.excludeCredentials) {
+    for (let i = 0; i < options.publicKey.excludeCredentials.length; i++) {
+      options.publicKey.excludeCredentials[i].id = decodeURLEncodedBase64(options.publicKey.excludeCredentials[i].id);
     }
+  }
 
-    navigator.credentials.create({
-      publicKey: makeCredentialOptions.publicKey
-    }).then(webauthnRegistered)
-      .catch((err) => {
-        if (!err) {
-          webAuthnError('unknown');
-          return;
-        }
-        webAuthnError('general', err.message);
-      });
-  }).fail((xhr) => {
-    if (xhr.status === 409) {
-      webAuthnError('duplicated');
-      return;
+  let credential;
+  try {
+    credential = await navigator.credentials.create({
+      publicKey: options.publicKey
+    });
+  } catch (err) {
+    if (err && err.message) {
+      webAuthnError('general', err.message);
+    } else {
+      webAuthnError('unknown', err);
```

Comment:
error is undefined here.

Reply:
`err` can never be undefined/falsy here, since it's provided by `catch`.

I think the old code is meant to process non-Error-type `err`, because in JS you can throw anything (well, I think if `navigator.credentials.create` is designed well, it should only throw Error types).

To be simple, use `err.toString()` is enough for the error message.

If you only 

```js
const p1 = new Promise((resolve, reject) => {
  throw 'string';
});
const p2 = new Promise((resolve, reject) => {
  throw new Error('error');
});

p1.catch((e) => console.error(e.toString()));
p2.catch((e) => console.error(e.toString()));
```

Output:

```
> string
> Error: error
```

Reply:
So `catch(e) {  webAuthnError('general', err.toString()); }` is enough